File: server/RestApiServer.Endpoints/ApiResponses/ApiResponse.cs
namespace RestApiServer.Endpoints.ApiResponses
{
    public enum ApiResponseType
    {
        Success,
        Failure,
        Error,
        File
    }

    public class ApiResponse<D>
    {
        public bool IsSuccessful { get; set; }
        public string Message { get; set; }
        public D? Data { get; set; }
        public string? ErrorCode { get; set; }
        public ApiResponseType ResponseType { get; set; }
        public int? StatusCode { get; set; }
        public Dictionary<string, string>? AdditionalInfo { get; set; }

        public ApiResponse(
            bool isSuccessful, 
            string message, 
            D? data, 
            ApiResponseType responseType,
            string? errorCode = null,
            int? statusCode = null,
            Dictionary<string, string>? additionalInfo = null)
        {
            IsSuccessful = isSuccessful;
            Message = message;
            Data = data;
            ResponseType = responseType;
            ErrorCode = errorCode;
            StatusCode = statusCode;
            AdditionalInfo = additionalInfo;
        }

        protected ApiResponse<D> AddInfo(string key, string value)
        {
            AdditionalInfo ??= new Dictionary<string, string>();
            AdditionalInfo[key] = value;
            return this;
        }
    }

    public class ApiResponseWithMetadata<D, M> : ApiResponse<D>
    {
        public M? Metadata { get; set; }

        public ApiResponseWithMetadata(
            bool isSuccessful, 
            string message, 
            D? data, 
            M? metadata,
            ApiResponseType responseType,
            string? errorCode = null,
            int? statusCode = null,
            Dictionary<string, string>? additionalInfo = null) 
            : base(isSuccessful, message, data, responseType, errorCode, statusCode, additionalInfo)
        {
            Metadata = metadata;
        }
    }
}

File: adminportal/src/ApiResponses/ApiResponse.ts
export enum ApiResponseType {
    Success = 'Success',
    Failure = 'Failure',
    Error = 'Error',
    File = 'File'
}

export class ApiResponse<D> {
    public additionalInfo?: Record<string, string>;

    constructor(
        public isSuccessful: boolean,
        public message: string,
        public data?: D,
        public responseType: ApiResponseType = ApiResponseType.Success,
        public errorCode?: string,
        public statusCode?: number,
        additionalInfo?: Record<string, string>
    ) {
        this.additionalInfo = additionalInfo;
    }

    public addInfo(key: string, value: string): ApiResponse<D> {
        if (!this.additionalInfo) {
            this.additionalInfo = {};
        }
        this.additionalInfo[key] = value;
        return this;
    }
}

export class ApiResponseWithMetadata<D, M> extends ApiResponse<D> {
    constructor(
        isSuccessful: boolean,
        message: string,
        data?: D,
        public metadata?: M,
        responseType: ApiResponseType = ApiResponseType.Success,
        errorCode?: string,
        statusCode?: number,
        additionalInfo?: Record<string, string>
    ) {
        super(isSuccessful, message, data, responseType, errorCode, statusCode, additionalInfo);
    }
}